homeassistant/components/fritzbox/__init__.py

```diff
@@ -84,8 +107,39 @@ async def async_setup_entry(hass, entry):
     except LoginError as err:
         raise ConfigEntryAuthFailed from err
 
-    hass.data.setdefault(DOMAIN, {CONF_CONNECTIONS: {}, CONF_DEVICES: set()})
-    hass.data[DOMAIN][CONF_CONNECTIONS][entry.entry_id] = fritz
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {
+        CONF_CONNECTIONS: fritz,
+        CONF_RETRY: False,
+    }
+
+    async def async_update_coordinator():
+        """Fetch all device data."""
+        try:
+            devices = await hass.async_add_executor_job(fritz.get_devices)
+        except requests.exceptions.HTTPError as ex:
+            if hass.data[DOMAIN][entry.entry_id][CONF_RETRY]:
+                raise ConfigEntryAuthFailed from ex
+            hass.data[DOMAIN][entry.entry_id][CONF_RETRY] = True
+            raise UpdateFailed(f"Fritzhome connection error: {ex}") from ex
+
+        data = {}
+        for device in devices:
+            await hass.async_add_executor_job(device.update)
```

Comment:
Since there are potentially many calls to the executor make a new function maybe `_update_fritz`, move everything out of `async_update_coordinator` into the new function and make `async_update_coordinator` call return await hass.async_add_executor_job(_update_fritz).  Then you only have one executor jump per update

Reply:
the update logic is now moved into a separate method